ACPI initial support on AArch64

michael2012z opened the pull request:
Partially fixes https://github.com/cloud-hypervisor/cloud-hypervisor/issues/2178

This is an early version of ACPI support on AArch64.

By implementing following tables on AArch64, guest VM can boot successfully:
MADT, DSDT, GTDT, SPCR, FADT, MCFG, XSDT,IORT.

Limitations:
- On AArch64, ACPI have to work with EFI. But before EFI ready, we are using a hack setting in VM cmdline `acpi_rsdp=`. This is only a temporary solution for test purpose.
- Other related functionalities (like hot-plug) have not been tested.

MrXinWang commented:
Also a little bit more information about the poweroff stuff from my side: Currently the only method to poweroff the guest is from the http API. For the poweroff using ACPI, it seems that we need to implement a dedicated GPIO controller (PL061) for AArch64 and using its pin 3 interrupt to trigger the poweroff (see: https://lists.gnu.org/archive/html/qemu-arm/2015-12/msg00169.html). 

I've already finished adding this PL061 device to the cloud-hypervisor, and currently I need to find out how to connect the GPIO interrupt with the ACPI event. Will raise another PR after finish the code.

rbradford commented:
It is cool that you've made so much progress on ACPI support. :rocket: 

> Also a little bit more information about the poweroff stuff from my side: Currently with this patch, the only method to poweroff the guest is from the http API. For the poweroff using ACPI, it seems that we need to implement a dedicated GPIO controller (PL061) for AArch64 and using its pin 3 interrupt to trigger the poweroff (see: https://lists.gnu.org/archive/html/qemu-arm/2015-12/msg00169.html).
> 
> I've already finished adding this PL061 device to the cloud-hypervisor, and currently I need to find out how to connect the GPIO interrupt with the ACPI event. Will raise another PR after finish the code.

What about using RESET_REG/SLEEP_CONTROL_REG: https://github.com/cloud-hypervisor/cloud-hypervisor/blob/master/vmm/src/acpi.rs#L146

These are currently port I/Os (so that it can be hardcoded in OVMF) but I see no reason why they couldn't be MMIO with the OVMF looking them up from this static table. You could try using MMIO for aarch64 with Linux and see if you can have the guest shutdown.

rbradford reviewed (commented):
I gave this a very quick scan through; i'll do a more detailed review later but I just wanted to give some positive feedback on how good this looks.

michael2012z commented:
Thank you @rbradford for cleaning the ACPI build on AArch64. That helped a lot for us to start.

I checked about "Reset Register" in ACPI spec, yes MMIO should work as well. Now we are trying to use `AcpiShutdownDevice` for AArch64.

michael2012z commented:
@rbradford I may not be able to reply quickly in following days, due to public holiday. Please leave your comments, I will handle when coming back.

The latest status on ACPI is: @MrXinWang and I are trouble shooting the problems in integration test, and investigating SHUTDOWN/RESET.

michael2012z commented:
As there is still work to do, I marked this PR draft. But any comments are welcome.

Status:
- Guest doesn't boot on 5.10.6 kernel with ACPI. I had been testing on 5.8. Now investigating the root cause.
- Debugging test cases. Temporarily I switched back to 5.8 in this PR to see the test case gap. Now following test cases are failing:
> 1. reboot
> 2. multi-cpus (only 1 CPU was brought up, could be related to missing MPIDR)
> 3. virtio-blk
> 4. virtio-fs

MrXinWang commented:
> As there is still work to do, I marked this PR draft. But any comments are welcome.
> 
> Status:
> 
> * Guest doesn't boot on 5.10.6 kernel with ACPI. I had been testing on 5.8. Now investigating the root cause.

For this issue, the root cause is from https://github.com/torvalds/linux/commit/1583052d111f8ea43f9954c5e749164fd2b954af#diff-6fdabe595f426a4d6462a3058737d00117dcf81f5fbb4e9fb1bdea32e9fc65baR269.

The latest kernel will disallow AML memory opregions to access kernel memory for AArch64 ACPI, which means if UEFI is not implemented, we will get a NULL pointer from `acpi_os_ioremap`. Therefore currently the AArch64 guest should only use any commit before this one, and in our case, v5.8 would be a good option.

michael2012z commented:
Hi, @rbradford , @sboeuf .  Now the integration test passed. The PR is ready for review. Before coming to details, I think we need to discuss which version of guest kernel to use.

Current situation:
- Now CLH is using kernel v5.10.6, but this PR cannot work with it. Our current test of ACPI is based on a hack based on 'acpi_rsdp=' parameter in cmdline. But since v5.9-rc1, this way has been disallowed, see commit https://github.com/torvalds/linux/commit/1583052d111f8ea43f9954c5e749164fd2b954af.
- There used to be a modified branch in CLH fork of linux kernel: clh-virtio-fs-virtio-iommu-5.8-rc4. It contains the backports of virtiofs with dax support. Now it is the best kernel tree this PR can work with.

So we could have 2 options:
1. Keep using v5.10.6, but do not enable ACPI in CI. We can add some documentation works about how to test ACPI on AArch64.
2. Switch back to modified v5.8, enable ACPI in CI. Stay to that version until EFI is ready.

What do you think?

michael2012z commented:
Status update. 

We are exploring the way to latest kernel version (ch-5.10.6). The question is how to put ACPI tables in ROM. We are trying to create a MMIO bus device for it.

rbradford commented:
Sorry it's taken so long for me to get a chance to look at this PR. With respect to the kernel version can we carry a patch to the 5.10 kernel fork to restore the ability to use `acpi_rsdp=` on the kernel command line. This would be preferrential to maintaining two separate forks of the kernel with different sets of patches to support the non-upstream parts (much reduced in 5.10.)

rbradford commented on acpi_tables/src/aml.rs:
```diff
@@ -543,6 +543,37 @@ impl Aml for IO {
+        bytes
+    }
+}
+
```
I think there are unit tests in this file. If there are it would be good to ensure that this is covered.

michael2012z replied:
I am on it.

michael2012z replied:
Done.

rbradford commented on vmm/src/acpi.rs:
```diff
@@ -192,20 +325,62 @@ pub fn create_acpi_tables(
         segment: 0,
         start: 0,
-        end: 0,
+        end: ((arch::layout::PCI_MMCONFIG_SIZE - 1) >> 20) as u8,
```
This looks like a good fix we should pull in

rbradford commented on vmm/src/acpi.rs:
```diff
@@ -105,23 +125,120 @@ pub fn create_dsdt_table(
+    iort
+}
+
 pub fn create_acpi_tables(
```
Rather than all the conditional code in this function, maybe we could split it into separate versions for aarch64 and x86_64 and extract common code (like FADT.)

michael2012z replied:
Please look at the latest 2 commits. The code of this file was largely refactored.

rbradford commented on vmm/src/cpu.rs:
```diff
@@ -1171,11 +1307,24 @@ struct CPU {
+            ],
+        )
+        .to_aml_bytes()
+    }
```
Ultimately I think you'll want the methods to support hotplug (or can you not hotplug CPUs on aarcht64?) so maybe you could reuse more of the version below but just exclude the _MAT field.

michael2012z replied:
Agree. Now I am reusing the code as much as possible. There should be some problems, errors were seen in guest kernel messages. More work is needed. But no test cases were broken. So I think we can live with these errors now, soon or later we will handle them.

rbradford commented on vmm/src/cpu.rs:
```diff
@@ -1435,10 +1585,14 @@ impl Aml for CpuManager {
             max_vcpus: self.config.max_vcpus,
         };
+        #[cfg(target_arch = "x86_64")]
         let mut cpu_data_inner: Vec<&dyn aml::Aml> = vec![&hid, &uid, &methods];
```
Ditto, I think it's fine to have the methods. We can just make them NOPs in the MMIO side or inside the methods themselves.

michael2012z replied:
Done.

rbradford reviewed (commented):
Well done for getting this far! :rocket:

michael2012z commented:
@rbradford Thanks for reviewing and discussing.

Before addressing all the comments, I added 2 more commits to refactor the code of building ACPI tables. I guess the code of `build_acpi_tables()` was difficult enough for you to read. Please have a look.

I wrote a test-only kernel patch at https://github.com/michael2012z/myLinux/commit/7a391a8b25512f3a7d19fe0097009502a154ab20. It removes the enhanced check on EFI and allows our test based on `acpi_rsdp=` hack. The latest PR is using this commit. Please have a look and pick it to CLH kernel fork.

We will continue fixing the CI test failure.

rbradford commented:
@michael2012z there is a new ch-5.10.6-aarch64-acpi-testing branch on cloud-hypervisor/linux

I think the ACPI table creation code looks better now, don't you?

michael2012z commented:
Status update: Serial device has not been enabled on ACPI, so all `serial` test cases failed. Debugging it.

michael2012z commented:
Progress update:
Serial device is ready on ACPI now. All relevant test cases passed.

There are still 2 test cases failed: `test_reboot` and `test_watchdog`, they are all related to `systemctl`.

michael2012z commented:
Found some clues about the problems with `systemd`:
- https://www.raspberrypi.org/forums/viewtopic.php?t=274452
- https://github.com/systemd/systemd/issues/662

Looks like they were caused by lack of RTC on ACPI. Investigating that.

michael2012z commented:
Status update:

On arm64 with ACPI, guest kernel obtains RTC information from UEFI. RTC device is not described in ACPI. See [here](https://patchwork.kernel.org/project/qemu-devel/patch/1452871445-4221-12-git-send-email-peter.maydell@linaro.org). So, before UEFI, we will not be able to fix the 2 RTC-related test case failures.

I will stop updating this PR for a while. In parallel with this PR, @jongwu is testing EDK2 with CLH on arm64. Once we can have UEFI + ACPI work together, I will come back to this PR for next steps then.

michael2012z commented on vmm/src/acpi.rs:
```diff
@@ -105,23 +124,252 @@ pub fn create_dsdt_table(
     dsdt
```
You (I) forgot to add checksum like:
`dsdt.update_checksum();`

This error slit from CI because checksum verification was ignored in early stage of kernel boot (see the message below). But it will definitely make problem later.

(From kernel boot messages:)
`[    0.000000] ACPI: Early table checksum verification disabled`

rbradford replied:
`.append_slice()` updates the checksum. If you force the checksum check on `acpi_force_table_verification` on the kernel command line it doesn't complain about it being invalid.

michael2012z replied:
Aha, yes we don't need another `update_checksum()` here. Thanks.

michael2012z reviewed (commented):
Checksum of DSDT was missing.

michael2012z commented:
Update:

In our local test, we can boot with ACPI + EDK2, in a very simple configuration (only 1 disk, no network). Now we can use the mainstream kernel, rather than the modified branch with a hack.

We updated integration test script to build EDK2, and test ACPI+EDK2 with `test_simple_launch` case, but it can not pass now. There are still problems to resolve:
1. EDK2 requires a flash space (now I give 128M) at the beginning of memory (where it used to be reserved area). CLH doesn't support flash, so I temporarily stole 128M from RAM for it. 
2. EDK2 checks the address of `PCIe high memory`. It requires the high memory at some where behind 4G. In CLH, that memory follows RAM. If I give RAM less than 3G (RAM starts from 1G) to the VM, the `PCIe high memory` will fall under 4G space, EDK2 fails.
3. When I added another disk (for cloudinit) or enable network, the virtio-pci device failed in initialization. This could be caused by the previous problems.

michael2012z commented:
Finally we manage to boot CLH with ACPI and UEFI. In this PR, a test case was added to test EDK2 + CLH (ACPI enabled).

A lot of thanks to @jongwu and @MrXinWang .

I think this PR is already too big. Next we will split it into several smaller ones, so the community can review easily.

Ping @rbradford @sboeuf .

MrXinWang commented on scripts/run_integration_tests_aarch64.sh:
```diff
@@ -8,9 +8,73 @@ export BUILD_TARGET=${BUILD_TARGET-aarch64-unknown-linux-gnu}
 mkdir -p "$WORKLOADS_DIR"
 
+build_edk2() {
+    EDK2_REPO="https://github.com/jongwu/edk2.git"
```
This should eventually point to either somewhere belongs to the CLH community (e.g. a forked branch) or the merged upstream repository

MrXinWang commented on vmm/Cargo.toml:
```diff
@@ -28,7 +28,7 @@ event_monitor = { path = "../event_monitor" }
 lazy_static = "1.4.0"
 libc = "0.2.94"
-linux-loader = { version = "0.3.0", features = ["elf", "bzimage", "pe"] }
+linux-loader = { git = "https://github.com/jongwu/linux-loader", branch = "uefi", version = "0.3.0", features = ["elf", "bzimage", "pe"] }
```
This should eventually point to either somewhere belongs to the CLH community (e.g. a forked branch) or the merged upstream repository (https://github.com/rust-vmm/linux-loader/pull/75)

michael2012z commented:
PR https://github.com/cloud-hypervisor/cloud-hypervisor/pull/2680 continues the remaining work of ACPI initial support on AArch64.
I will close this PR.
